### PR TITLE
[Perf] Optimize `CutlassFP8ScaledMMLinearKernel` when padding needed by pre-weight processing, 13.5% TTFT improvement

### PR DESCRIPTION
**Pull request by yewentao256**
## Purpose

We quantize the weight each time we do a forward, this could be optimized through pre-weight processing

## Test

`VLLM_DISABLED_KERNELS=MarlinFP8ScaledMMLinearKernel,FlashInferFP8ScaledMMLinearKernel VLLM_LOGGING_LEVEL=INFO vllm serve RedHatAI/DeepSeek-Coder-V2-Lite-Instruct-FP8 --tensor-parallel-size 8 --host 127.0.0.1 --port 9256`

### Acc

`lm_eval --model local-completions --model_args "base_url=http://127.0.0.1:9256/v1/completions,model=$MODEL,num_concurrent=1024" --tasks gsm8k`

```bash
# Now
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7460|±  |0.0120|
|     |       |strict-match    |     5|exact_match|↑  |0.7293|±  |0.0122|
# main
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.7460|±  |0.0120|
|     |       |strict-match    |     5|exact_match|↑  |0.7263|±  |0.0123|

```

### Perf

```bash
vllm bench serve --backend vllm --model RedHatAI/DeepSeek-Coder-V2-Lite-Instruct-FP8 --host 127.0.0.1 --port 9256 --dataset-name random --random-input-len 16 --random-output-len 1024 --request-rate inf --max-concurrency 32 --num-prompts 256 --num-warmups 32

# Now

============ Serving Benchmark Result ============
Successful requests:                     256       
Failed requests:                         0         
Maximum request concurrency:             32        
Benchmark duration (s):                  71.94     
Total input tokens:                      4096      
Total generated tokens:                  262144    
Request throughput (req/s):              3.56      
Output token throughput (tok/s):         3644.00   
Peak output token throughput (tok/s):    3807.00   
Peak concurrent requests:                64.00     
Total token throughput (tok/s):          3700.93   
---------------Time to First Token----------------
Mean TTFT (ms):                          39.04     
Median TTFT (ms):                        34.56     
P99 TTFT (ms):                           69.29     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          8.75      
Median TPOT (ms):                        8.76      
P99 TPOT (ms):                           8.77      
---------------Inter-token Latency----------------
Mean ITL (ms):                           8.75      
Median ITL (ms):                         8.74      
P99 ITL (ms):                            9.58      
==================================================

# main
============ Serving Benchmark Result ============
Successful requests:                     256       
Failed requests:                         0         
Maximum request concurrency:             32        
Benchmark duration (s):                  72.14     
Total input tokens:                      4096      
Total generated tokens:                  262144    
Request throughput (req/s):              3.55      
Output token throughput (tok/s):         3633.85   
Peak output token throughput (tok/s):    3732.00   
Peak concurrent requests:                64.00     
Total token throughput (tok/s):          3690.63   
---------------Time to First Token----------------
Mean TTFT (ms):                          44.32     
Median TTFT (ms):                        41.21     
P99 TTFT (ms):                           79.10     
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          8.76      
Median TPOT (ms):                        8.76      
P99 TPOT (ms):                           8.81      
---------------Inter-token Latency----------------
Mean ITL (ms):                           8.76      
Median ITL (ms):                         8.77      
P99 ITL (ms):                            9.47      
==================================================
```

### CODE REVIEW

**Review by yewentao256 — COMMENTED** [review]
CC @MatthewBonanni if you are available, thanks!

File: vllm/model_executor/kernels/linear/scaled_mm/cutlass.py [hunk 1]
```diff
@@ -187,39 +216,24 @@ def apply_scaled_mm(
...
-                # Bs is per-tensor (numel==1) or per-channel (numel==N)
-                # in this kernel class — never 2D block-wise.
-                if Bs.numel() > 1:
-                    Bs = self._pad_to_alignment(
-                        Bs.view(-1), dim=0, alignment=16, value=1.0
-                    )
-                    if Bs.dim() == 1 and B.shape[1] > 1:
-                        Bs = Bs.view(-1, 1)
+        padded_k, padded_n = B.shape
+        output_size = getattr(self, "logical_output_size", output_shape[-1])
```
Comment (MatthewBonanni) [inline vllm/model_executor/kernels/linear/scaled_mm/cutlass.py]: Could we instead initialize `self.logical_output_size = None` in the constructor and then set it in `process_weights_after_loading` (before early return)?

Then we can just use `self.logical_output_size`

File: vllm/model_executor/kernels/linear/scaled_mm/cutlass.py [hunk 2]
```diff
@@ -176,6 +176,35 @@ def _pad_to_alignment(
...
         return torch.nn.functional.pad(x, pad_spec, value=value)
 
+    def process_weights_after_loading(self, layer: torch.nn.Module) -> None:
+        weight_name, weight_scale_name, _, _ = self.layer_param_names
+        weight = getattr(layer, weight_name)
+
+        # keep the logical output width so runtime can slice away static padding.
+        self.logical_output_size = getattr(
+            layer, "output_size_per_partition", weight.shape[1]
+        )
```
Comment (MatthewBonanni) [inline vllm/model_executor/kernels/linear/scaled_mm/cutlass.py]:
```suggestion
        self.logical_output_size = weight.shape[1]
```

File: vllm/model_executor/kernels/linear/scaled_mm/cutlass.py (hunk 1 above)
Comment (yewentao256) [reply vllm/model_executor/kernels/linear/scaled_mm/cutlass.py]: Solved

**Review by MatthewBonanni — APPROVED** [review]
LGTM, thanks!

**Review by mgoin — APPROVED** [review]

**Comment by njhill** [comment]
This had failed tests which look like they may have been caused by the PR and now failing on main, maybe we should revert for now?

 e.g.https://buildkite.com/vllm/ci/builds/66955/list?jid=019e419c-65ea-4dd3-9e24-41d776f5c873&tab=output

**Comment by yewentao256** [comment]
> This had failed tests which look like they may have been caused by the PR and now failing on main, maybe we should revert for now?
> 
> e.g.https://buildkite.com/vllm/ci/builds/66955/list?jid=019e419c-65ea-4dd3-9e24-41d776f5c873&tab=output

Let me fix that
